Sources/xcproj/Referenceable.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  Referenceable.swift
+//  xcprojPackageDescription
+//
+//  Created by rmalik on 11/8/17.
+//
+
+import Foundation
+
+public protocol Referenceable {
+    var reference: String { get }
+}
+
+extension Array where Element: Referenceable {
+
+    public var references: [String] {
+        return map { $0.reference }
+    }
+
+    public func contains(reference: String) -> Bool {
+        return contains { $0.reference == reference }
+    }
+
+    public func getReference(_ reference: String) -> Element? {
+        return first { $0.reference == reference }
+    }
+}
+
+extension Dictionary where Value: Referenceable {
+    public var references: [String] {
+        return self.values.map { $0.reference }
+    }
+
+    public func contains(reference: Key) -> Bool {
+        return self[reference] != nil
+    }
+
+    public func getReference(_ reference: Key) -> Value? {
+        return self[reference]
+    }
+}
+
+public struct ReferenceableCollection<T: Referenceable & Equatable>: Equatable, ExpressibleByArrayLiteral {
```

Comment:
@rahul-malik What benefits do we get out of using this struct as a opposed to a Dictionary?

Reply:
@yonaskolb - The benefits are largely that you have an abstraction around the internal data structure and the mutations are only additive (through `add(_)`). `PBXProj` consumes many arrays and this simplifies the conversion to the internal dictionary representation.

For me this is more about encapsulation and constraining the data type to match it's usage. 

Thoughts?

Reply:
I'm not quite convinced it's worth it. It seems to just be mirroring a Dictionary API with some extra helpers (that could still be added as a constrained extension on a dictionary). We also lose out on any of the new Swift 4 dictionary apis, and  - in the current form - any way to remove objects. 
Personally I would say these could all just be `[String: PBXFileReference]` like you had before.

Reply:
@yonaskolb - I could certainly look into changing it back. For use case I'm never removing objects from the projects but if that is a use-case we support then that makes sense.

@pepibumur - Thoughts?

Reply:
I get your point about abstraction @rahul-malik but comparing the benefits of having such abstraction, and the downside of having to maintain the interface and keep up with `Foundation` changes I think it's not that worth having such abstraction. I would see a benefit in being able to optimize stuff keeping the interface intact but I'm not sure how further we can go with optimization at the collection level once we move from arrays to dictionaries. In case we need to change the collections types in the future, we can always extend them to match the interface that we are currently exposing.

Is there any other benefit that I might be missing @rahul-malik?

Reply:
@rahul-malik yeah we need to support removing as well, as that comes as part of modifying a project. I understand now you wanted to restrict that, hence the encapsulation